File: first_and_last_in_sorted_array.py
class Solution(object):
    def __init__(self):
        self.firstIndex = -1

    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        lo, hi = 0, len(nums) - 1

        findFirstIndex = self.binarySearch(nums, 0, hi, target)
        if findFirstIndex == -1: return [-1, -1]
        findSecondIndex = self.secondBinarySearch(nums, self.firstIndex, len(nums) - 1, target)
        return [findFirstIndex, findSecondIndex]

    def binarySearch(self, nums, lo, hi, target):
        foundIndex = -1
        while lo <= hi:
            mid = lo + (hi - lo) // 2
            if nums[mid] == target:

                if (mid == 0 or mid == hi or nums[mid - 1] < nums[mid]):
                    if self.firstIndex == -1: self.firstIndex = mid
                    foundIndex = mid
            if target <= nums[mid]:
                hi = mid - 1
            else:
                lo = mid + 1
        return foundIndex

    def secondBinarySearch(self, nums, lo, hi, target):
        foundIndex = -1
        while lo <= hi:
            mid = lo + (hi - lo) // 2
            if nums[mid] == target:

                if (mid == 0 or mid == hi or nums[mid] < nums[mid + 1]):
                    foundIndex = mid
            if target >= nums[mid]:
                lo = mid + 1
            else:
                hi = mid - 1

        return foundIndex

File: FirstAndLastInSortedArray.java
// sc - O(1)
class FirstAndLastInSortedArray {
    public int[] searchRange(int[] nums, int target) {

        if(nums.length == 0 || nums.length == 0)
            return new int[]{-1, -1};

        int n = nums.length;
        if(nums[n-1] < target || nums[0] > target) return new int[]{-1, -1};

        int left = binarySearchLeft(nums, target);
        if(left == -1) return new int[]{-1, -1};
        int right = binarySearchRight(nums, left, n-1, target);
        if(right == -1) return new int[]{-1, -1};
        System.out.println(left + " " + right);

        return new int[]{left, right};
    }

    public int binarySearchLeft(int[] nums, int target) {
        int low = 0, high = nums.length-1;

        while (low <= high) {
            int mid = low + (high-low) / 2;

            if(nums[mid] == target) {
                if(mid == high || nums[mid] > nums[mid-1]) {
                    return mid;
                } else {
                    high = mid - 1;
                }
            } else if(nums[mid] > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return -1;
    }

    public int binarySearchRight(int[] nums, int low, int high, int target) {
        // int low = 0, high = nums.length-1;

        while (low <= high) {
            int mid = low + (high-low) / 2;

            if(nums[mid] == target) {
                if(nums[mid] > nums[mid+1]) {
                    return mid;
                } else {
                    low = mid + 1;
                }
            } else if(nums[mid] > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return -1;
    }
}